File: engine/src/main/java/org/epdb/engine/volcano/Projection.java
package org.epdb.engine.volcano;

import java.util.List;
import java.util.Set;

import org.epdb.engine.columntypes.ColumnValue;
import org.epdb.engine.dto.Tuple;

public record Projection(
        Operator childOperator,
        Set<Integer> projectionColumns
) implements Operator {

    @Override
    public void open() {
        this.childOperator.open();
    }

    @Override
    public Tuple next() {
        var tuple = childOperator.next();

        if (tuple == null) {
            return null;
        }

        ColumnValue[] values = new ColumnValue[this.projectionColumns.size()];
        int k = 0;
        for (var i = 0; i < tuple.getValues().size(); i++) {
            if (this.projectionColumns.contains(i)) {
                values[k++] = tuple.getValueAtIndex(i);
            }
        }

        return new Tuple(List.of(values));
    }

    @Override
    public void close() {
        this.childOperator.close();
    }
}

File: engine/src/main/kotlin/org/epdb/engine/databaseoperator/Projection.kt
package org.epdb.engine.databaseoperator

import org.epdb.engine.dto.Tuple

data class Projection(
    private val childOperator: Operator,
    private val projectionColumns: Set<Int>
) : Operator {

    override fun open() {
        this.childOperator.open()
    }

    override fun next(): Tuple? {
        val tuple = childOperator.next() ?: return null

        return Tuple(
            values = tuple.values.filterIndexed { index, _ -> projectionColumns.contains(index) }
        )
    }

    override fun close() {
        this.childOperator.close()
    }
}
